### DISCUSSION TITLE
Wrong PSRam tasmota32s3-qio_opi (multiple values)

### DISCUSSION THREAD

**Discussion by valdus2** [arendst/Tasmota#21607] [Q&A]
Until version 13.4.1.2 Core/SDK Version: 3_0_0/5.1.3.240430, PSRam correctly detected PS-RAM Memory: **8192 KB**.

Using tasmota32s3-qio_opi.bin 3_0_1/5.1.4.240602 from https://github.com/tasmota/install/tree/firmware/firmware/unofficial
PS-RAM Memory: **5568 KB**.

Using Tasmota32-S3 (QIO Flash/OPI PSRAM) fromWeb Installer
PS-RAM Memory: **5314 KB**.

Compiling released 14.1.0 Core/SDK Version: 3_0_0/5.1.4.240529 (`platformio run -e tasmota32s3-qio_opi-all`)
PS-RAM Memory: **6016 KB**.

Compiling development 14.1.0.1 Core/SDK Version: 3_0_1/5.1.4.240602 (`platformio run -e tasmota32s3-qio_opi-all`)
PS-RAM Memory: **6016 KB**. (with BLE)
PS-RAM Memory: **6144 KB**. (No BLE)
 
 
 

Reinstalling 13.4.1.2 again correctly identified PS-RAM Memory: **8192 KB**.
(module: ESP32-S3-DevKitC-1-N8R8)

What can I do to get back the correct PSRam with latest versions?
Thank you in advance.

**Comment by sfromis** [comment]
There is nothing "wrong" there. What happens is that there is an optimization when having a large amount on octal PSRAM with ESP32-S3, copying Tasmota from flash to PSRAM for enhanced speed. Seems to provide improved performance with at least some types of display operation.

Do you have a use case for using all that PSRAM for "better purposes"?

**Comment by valdus2** [comment]
But the total identified PSRam, shouldn't be the correct one, 8192KB? As much as the chip has.
After that, how it is used by tasmota should be reflected in free PSRam.

So you say that the ram missing is occupied by tasmota image? But flash remains the same occupied by the flashed image. 

And why there are so many different values considering different images?
If there is an optimization, shouldn't always return the same value as PSRam?

**Reply by sfromis** [reply]
What you now call "identified" is just what's available to Tasmota. As Tasmota was *already* copied to PSRAM, it cannot see or use more than what was left after boot. And the "free" PSRAM is what's free from the chunk that Tasmota can allocate within.

And I said *copied*, not moved. As PSRAM goes away when the device is powered off, that cannot be a substitute for always having the Tasmota image in flash. And then possibly copied to PSRAM for enhanced performance, as an optimization.

Different sizes are due to you using Tasmota images of different sizes.

Not clear what you mean with that "shouldn't" in the last line. As mentioned, the copying happens early in boot, meaning that Tasmota never sees the PSRAM it was copied to. But again, when you see "PS-RAM Memory 5568 KB", that's not about what the chip has, but about what's at the disposition of Tasmota after boot.

**Comment by valdus2** [comment]
> Do you have a use case for using all that PSRAM for "better purposes"?

I was just checking two of these modules with Haspmota and they use about 5mb of PSRam
So if I used for example Tasmota32-S3 (QIO Flash/OPI PSRAM) fromWeb Installer, then I couldn't use them...

So, is there a build flag, a setoption or whatever, to have the whole psram at my disposal as before ?
(Choosing if we want this optimization)

This is really very important especially when everything runs fine without this optimization, as in 13.4.1.2, 
but what we really need, is the whole PSRam.

**Reply by sfromis** [reply]
As I'm not deep into details of the build setup, IDK where the option to copy Tasmota code from flash to ESP32-S3 octal PSRAM is selected, @Jason2866

**Comment by Jason2866** [comment]
It is a compile time option for OPI PSRAM when compiling the Arduino libs. There is no way to change either in Arduino nor in Tasmota code. Since this option fixes Screen flickering with some displays it will stay.

More correct it is not Tasmota code copied into PSRAM, low level routines from IDF stored in flash are copied to PSRAM to avoid access to Flash which would causes screen flicker issues.
The amount of shown PSRAM is what we get from the System API functions. It shows the amount of useable PSRAM for the programm. You are welcome to provide a PR which shows additionally the hardware size of the PSRAM.

**Reply by sfromis** [reply]
I'm surprised that "low level routines from IDF stored in flash are copied to PSRAM" just "happens" to be very close to the Tasmota code size.

**Comment by valdus2** [comment]
Yes but as I stated in my above post I Really need PSRam!
As I have no flickering devices and everything was working beautifully before,
now you're telling me what? Buy new modules ???

As this is an option as you said there must be a way of rolling back this behavior.
And no, my problem is not showing the hardware size of the PSRAM but having it all for use!

**Comment by Jason2866** [comment]
Compile your own Arduino framework where you remove the OPI speed optimizations and compile with this framework Tasmota. As said, this setting will stay, since it fixes screen flicker for many. Needing more than 5MB  of PSRAM is not common.

**Comment by valdus2** [comment]
@Jason2866
This can't be an Accepted Answer to a problem.
This is an Accepted Explanation of the problem.
And I'm not an Arduino framework expert so this isn't something that I could do...

#### I've just lost **40%** of my PSRam just by updating, if that's not a problem, then I wonder what is...!

Of cpurse, you understand that this has nothing to do with you in personally.

**Reply by sfromis** [reply]
Of course, you could also choose to stay with the previous Tasmota version. For most purposes, there is not much value in "keeping up with the numbers".

**Reply by valdus2** [reply]
Is there a way to compile in gitpod using older versions of tasmota, either released or development?

**Comment by Jason2866** [comment]
I see nowhere a decrease of 40% of PSRAM. Correct we use a part of PSRAM now to fix a major issue.
You can checkout every commit of Tasmota and compile this version. Using a older Tasmota Arduino framework with actual Tasmota is not tested nor supported. May work or not.

**Reply by valdus2** [reply]
> I see nowhere a decrease of 40% of PSRAM


In my first post :

Until version 13.4.1.2 Core/SDK Version: 3_0_0/5.1.3.240430, PS-RAM Memory: 8192 KB.

Using Tasmota32-S3 (QIO Flash/OPI PSRAM) from Web Installer: 
PS-RAM Memory: 5314 KB. After boot without using anything, Free 5100KB. 
 

That's why I'm complaining...

**Reply by Jason2866** [reply]
Yes, the S3 QIO OPI build from https://github.com/tasmota/install/tree/firmware/firmware/unofficial is a "fat" build with a lot enabled. Never said that's an optimal one for Haspmota use.
It is in "unofficial" section ;-)

**Reply by valdus2** [reply]
With the best scenario I loose 30%. From to 8192 to 6016 KB. (Lower free after booting.)
And that involves no domoticz, no knx, no matter, and I don't remember what else -no-...

**Reply by sfromis** [reply]
Did you also disable larger features like LVGL/HASPmota and Webcam? It should be feasible to get lower than your number suggests, but still a lot of PSRAM taken up when this matches the size of the binary.

**Reply by valdus2** [reply]
Yes, of course webcam, but no LVGL/HASPmota...That's the necessity for PSRam, for Haspmota...

**Comment by Jason2866** [comment]
@sfromis This exactly is done 
```
CONFIG_LCD_RGB_ISR_IRAM_SAFE=y
 CONFIG_GDMA_CTRL_FUNC_IN_IRAM=y
 # bounce buffer mode relies on GDMA EOF interrupt to be service-able
 CONFIG_GDMA_ISR_IRAM_SAFE=y
 # Enable the XIP-PSRAM feature, so the ext-mem cache won't be disabled when SPI1 is operating the main flash
 CONFIG_SPIRAM_FETCH_INSTRUCTIONS=y
 CONFIG_SPIRAM_RODATA=y
```
https://docs.espressif.com/projects/esp-idf/en/v5.1.4/esp32s3/api-reference/kconfig.html#config-spiram-fetch-instructions
https://docs.espressif.com/projects/esp-idf/en/v5.1.4/esp32s3/api-reference/kconfig.html#config-spiram-rodata

**Reply by sfromis** [reply]
Well, as the reduction in available PSRAM is very close to the Tasmota binary size, it does look like it is copying all of the code from the binary to PSRAM, a lot more than "low level routines from IDF".

**Reply by valdus2** [reply]
Program Size | 2249 KB
PSRam used | 2176 KB
That's almost the same....
